test/IECoreScene/ShaderNetworkAlgoTest.py

```diff
@@ -263,6 +263,100 @@ def testArnoldComponentConnectionsNotConverted( self ) :
+		pack.parameters["in1"] = IECore.FloatData( 0.0 )
+		network.addShader( "pack", pack )
+		network.addConnection( IECoreScene.ShaderNetwork.Connection(
+			IECoreScene.ShaderNetwork.Parameter( "source1", "out" ),
+			IECoreScene.ShaderNetwork.Parameter( "pack", "in1" )
+		) )
+		network.addConnection( IECoreScene.ShaderNetwork.Connection(
+			IECoreScene.ShaderNetwork.Parameter( "pack", "out" ),
+			IECoreScene.ShaderNetwork.Parameter( "dest", "d" )
+		) )
+		# Manual swizzle ( should not be affected by adapters )
+		swizzle = IECoreScene.Shader( "MaterialX/mx_swizzle_color_float", "osl:shader" )
+		swizzle.parameters["in"] = IECore.Color3fData( imath.Color3f( 0.0 ) )
+		swizzle.parameters["channels"] = IECore.StringData( "R" )
+		network.addShader( "swizzle", swizzle )
+		network.addConnection( IECoreScene.ShaderNetwork.Connection(
+			IECoreScene.ShaderNetwork.Parameter( "source2", "out" ),
+			IECoreScene.ShaderNetwork.Parameter( "swizzle", "in" )
+		) )
+		network.addConnection( IECoreScene.ShaderNetwork.Connection(
+			IECoreScene.ShaderNetwork.Parameter( "swizzle", "out" ),
+			IECoreScene.ShaderNetwork.Parameter( "dest", "e" )
+		) )
+
+		converted = network.copy()
+		IECoreScene.ShaderNetworkAlgo.addComponentConnectionAdapters( converted )
+		prefixMismatch = network.copy()
+		IECoreScene.ShaderNetworkAlgo.addComponentConnectionAdapters( prefixMismatch, "doesNotExist" )
+		prefixMatch = network.copy()
+		IECoreScene.ShaderNetworkAlgo.addComponentConnectionAdapters( prefixMatch, "ai" )
+		convertBack = converted.copy()
+		IECoreScene.ShaderNetworkAlgo.removeComponentConnectionAdapters( convertBack )
+
+		connectionDict = dict( [ ( i.destination.name, i ) for i in converted.inputConnections( "dest" ) ] )
+		self.assertEqual( converted.getShader( connectionDict["a"].source.shader ).name, "MaterialX/mx_pack_color" )
+		self.assertEqual( converted.getShader( connectionDict["b"].source.shader ).name, "MaterialX/mx_pack_color" )
+		self.assertEqual( converted.getShader( connectionDict["c"].source.shader ).name, "MaterialX/mx_swizzle_color_float" )
+
+		self.assertEqual( IECoreScene.ShaderNetworkAlgo.componentConnectionAdapterLabel(), IECore.InternedString( "cortex_autoAdapter" ) )
+		self.assertEqual( converted.getShader( connectionDict["c"].source.shader ).blindData(), IECore.CompoundData( { "cortex_autoAdapter" : True } ) )
+
+		# With a prefix that doesn't match, nothing happens
+		self.assertEqual( prefixMismatch, network )
+		# With a prefix that matches, everything gets adapted
+		self.assertEqual( prefixMatch, converted )
+
+		# Check round trip
+		self.assertEqual( convertBack, network )
+
+		# Check that the shaders we don't modify are reused when the network is copied ( the addresses are the same )
+		self.assertEqual( str( network.getShader( "source1" ) ), str( converted.getShader( "source1" ) ) )
+		self.assertEqual( str( network.getShader( "dest" ) ), str( converted.getShader( "dest" ) ) )
```

Comment:
I don't understand this. For RefCounted objects, `a == b` or `a.isSame( b )` would check identity accurately. And I think it would fail here, because `ShaderNetwork.getShader()` returns a copy to Python (so you can't accidentally modify the thing that was const in C++). Which makes me thing that the string comparison is probably bogus?

In several places in Gaffer (e.g. `TypedObjectPlug.getValue()`), we use an optional `_copy` argument to control the behaviour of functions like `getShader`, so perhaps we should add that and then do this?

```
self.assertEqual( network.getShader( "source1", _copy = False ), converted.getShader( "source1", _copy = False ) )
```

Reply:
Oh, fair.  I didn't really think too much about this one, just saw that the string representation contained an address, and it matched, which seemed right, and figured it was good.   I guess you're saying that the Python wrapper is probably allocating a new object at this address, and because it immediately is discarded, the next copy object ends up at the same address?

If this was independent code, it seems like it would be healthier to just add a "getAddress" to be used for debugging, rather than "_copy = False", which seems more confusing and harmful, if only needed for debugging, but I guess "_copy = False" is the convention.

At the end, you mean I should use `self.assertTrue( network.getShader( ... ).isSame( network.getShader( ... ) )`, right?  The _copy wouldn't do anything if we were just doing value comparison?

Reply:
> At the end, you mean I should use self.assertTrue( network.getShader( ... ).isSame( network.getShader( ... ) ), right? The _copy wouldn't do anything if we were just doing value comparison?

Oh, yep, that looks good. I was thinking `assertEqual()` would do the job because the RefCounted bindings compare by pointer, but the Object bindings override that to compare values. So yep, `assertTrue( a.isSame( b ) )` is it.